### PR TITLE
[SPIR-V] Add support for empty string

### LINKED ISSUE #4360: [SPIR-V] Shader DebugInfo (-fspv-debug=vulkan-with-source) causes LLVM Assert (Debug) and Crash (Release) with string literals
I built DXC from source from 2dc067b561f17d09d8012a1ded05bf0f6253fea5 to test out shader debugging in our engine.  I'm seeing a crash when I invoke compilation via the API and an Assert in Debug.  I stripped it down to a simple repro:

```
// This string variable causes the crash - remove it to make crash go away
string g_renderState_FillMode < string arg1 = "WIREFRAME" ; > = "" ;  
struct PS_OUTPUT 
{ 
    float4 vColor0 : SV_Target0 ; 
} ; 

PS_OUTPUT MainPs (  ) 
{ 
    PS_OUTPUT o ; 
    o . vColor0 = float4( 0, 0, 0, 1 );
    return o ; 
} 
```
Compile with this:
```
dxc.exe -EMainPs -Wno-effects-syntax -T ps_6_0 -fspv-target-env=vulkan1.1 -fspv-debug=vulkan-with-source -spirv test.hlsl
```
I get:
```
Internal compiler error: LLVM Assert
```

When I try it via the API, I see this in the debugger:

```
 	KernelBase.dll!00007ff960904f69()	Unknown
>	dxcompiler.dll!llvm_assert(const char * _Message, const char * _File, unsigned int _Line, const char * _Function) Line 23	C++
 	dxcompiler.dll!llvm::DenseMapBase<llvm::DenseMap<llvm::StringRef,clang::spirv::SpirvString *,clang::spirv::StringMapInfo,llvm::detail::DenseMapPair<llvm::StringRef,clang::spirv::SpirvString *>>,llvm::StringRef,clang::spirv::SpirvString *,clang::spirv::StringMapInfo,llvm::detail::DenseMapPair<llvm::StringRef,clang::spirv::SpirvString *>>::LookupBucketFor<llvm::StringRef>(const llvm::StringRef & Val, const llvm::detail::DenseMapPair<llvm::StringRef,clang::spirv::SpirvString *> * & FoundBucket) Line 482	C++
 	dxcompiler.dll!llvm::DenseMapBase<llvm::DenseMap<llvm::StringRef,clang::spirv::SpirvString *,clang::spirv::StringMapInfo,llvm::detail::DenseMapPair<llvm::StringRef,clang::spirv::SpirvString *>>,llvm::StringRef,clang::spirv::SpirvString *,clang::spirv::StringMapInfo,llvm::detail::DenseMapPair<llvm::StringRef,clang::spirv::SpirvString *>>::LookupBucketFor<llvm::StringRef>(const llvm::StringRef & Val, llvm::detail::DenseMapPair<llvm::StringRef,clang::spirv::SpirvString *> * & FoundBucket) Line 521	C++
 	dxcompiler.dll!llvm::DenseMapBase<llvm::DenseMap<llvm::StringRef,clang::spirv::SpirvString *,clang::spirv::StringMapInfo,llvm::detail::DenseMapPair<llvm::StringRef,clang::spirv::SpirvString *>>,llvm::StringRef,clang::spirv::SpirvString *,clang::spirv::StringMapInfo,llvm::detail::DenseMapPair<llvm::StringRef,clang::spirv::SpirvString *>>::InsertIntoBucketImpl(const llvm::StringRef & Key, llvm::detail::DenseMapPair<llvm::StringRef,clang::spirv::SpirvString *> * TheBucket) Line 443	C++
 	dxcompiler.dll!llvm::DenseMapBase<llvm::DenseMap<llvm::StringRef,clang::spirv::SpirvString *,clang::spirv::StringMapInfo,llvm::detail::DenseMapPair<llvm::StringRef,clang::spirv::SpirvString *>>,llvm::StringRef,clang::spirv::SpirvString *,clang::spirv::StringMapInfo,llvm::detail::DenseMapPair<llvm::StringRef,clang::spirv::SpirvString *>>::InsertIntoBucket(llvm::StringRef && Key, clang::spirv::SpirvString * && Value, llvm::detail::DenseMapPair<llvm::StringRef,clang::spirv::SpirvString *> * TheBucket) Line 419	C++
 	dxcompiler.dll!llvm::DenseMapBase<llvm::DenseMap<llvm::StringRef,clang::spirv::SpirvString *,clang::spirv::StringMapInfo,llvm::detail::DenseMapPair<llvm::StringRef,clang::spirv::SpirvString *>>,llvm::StringRef,clang::spirv::SpirvString *,clang::spirv::StringMapInfo,llvm::detail::DenseMapPair<llvm::StringRef,clang::spirv::SpirvString *>>::FindAndConstruct(llvm::StringRef && Key) Line 243	C++
 	dxcompiler.dll!llvm::DenseMapBase<llvm::DenseMap<llvm::StringRef,clang::spirv::SpirvString *,clang::spirv::StringMapInfo,llvm::detail::DenseMapPair<llvm::StringRef,clang::spirv::SpirvString *>>,llvm::StringRef,clang::spirv::SpirvString *,clang::spirv::StringMapInfo,llvm::detail::DenseMapPair<llvm::StringRef,clang::spirv::SpirvString *>>::operator[](llvm::StringRef && Key) Line 246	C++
 	dxcompiler.dll!clang::spirv::SpirvBuilder::getString(llvm::StringRef str) Line 1645	C++
 	dxcompiler.dll!clang::spirv::DeclResultIdMapper::createOrUpdateStringVar(const clang::VarDecl * var) Line 1105	C++
 	dxcompiler.dll!clang::spirv::SpirvEmitter::doVarDecl(const clang::VarDecl * decl) Line 1556	C++
 	dxcompiler.dll!clang::spirv::SpirvEmitter::doDecl(const clang::Decl * decl) Line 813	C++
 	dxcompiler.dll!clang::spirv::SpirvEmitter::HandleTranslationUnit(clang::ASTContext & context) Line 676	C++
 	dxcompiler.dll!clang::ParseAST(clang::Sema & S, bool PrintStats, bool SkipFunctionBodies) Line 162	C++
 	dxcompiler.dll!clang::ASTFrontendAction::ExecuteAction() Line 556	C++
 	dxcompiler.dll!clang::FrontendAction::Execute() Line 468	C++
 	dxcompiler.dll!DxcCompiler::Compile(const DxcBuffer * pSource, const wchar_t * * pArguments, unsigned int argCount, IDxcIncludeHandler * pIncludeHandler, const _GUID & riid, void * * ppResult) Line 1031	C++
```
When I use a Release build of DXC I get a crash with the following callstack:
```
D:\Users\Dan\src\DirectXShaderCompiler\include\llvm/ADT/DenseMap.h(484)
Func:	llvm::DenseMapBase<class llvm::DenseMap<class llvm::StringRef,class clang::spirv::SpirvString *,struct clang::spirv::StringMapInfo,struct llvm::detail::DenseMapPair<class llvm::StringRef,class clang::spirv::SpirvString *> >,class llvm::StringRef,class clang::spirv::SpirvString *,struct clang::spirv::StringMapInfo,struct llvm::detail::DenseMapPair<class llvm::StringRef,class clang::spirv::SpirvString *> >::LookupBucketFor
Exception thrown at 0x00007FF960904F69 (KernelBase.dll) in ...exe: 0xE0000001.
Unhandled exception at 0x00007FF960904F69 (KernelBase.dll) in ...exe: 0xE0000001.

>	dxcompiler.dll!llvm::DenseMapBase<llvm::DenseMap<llvm::StringRef,clang::spirv::SpirvString *,clang::spirv::StringMapInfo,llvm::detail::DenseMapPair<llvm::StringRef,clang::spirv::SpirvString *>>,llvm::StringRef,clang::spirv::SpirvString *,clang::spirv::StringMapInfo,llvm::detail::DenseMapPair<llvm::StringRef,clang::spirv::SpirvString *>>::LookupBucketFor<llvm::StringRef>(const llvm::StringRef & Val, const llvm::detail::DenseMapPair<llvm::StringRef,clang::spirv::SpirvString *> * & FoundBucket) Line 486	C++
 	[Inline Frame] dxcompiler.dll!llvm::DenseMapBase<llvm::DenseMap<llvm::StringRef,clang::spirv::SpirvString *,clang::spirv::StringMapInfo,llvm::detail::DenseMapPair<llvm::StringRef,clang::spirv::SpirvString *>>,llvm::StringRef,clang::spirv::SpirvString *,clang::spirv::StringMapInfo,llvm::detail::DenseMapPair<llvm::StringRef,clang::spirv::SpirvString *>>::InsertIntoBucketImpl(const llvm::StringRef & TheBucket, llvm::detail::DenseMapPair<llvm::StringRef,clang::spirv::SpirvString *> *) Line 449	C++
 	[Inline Frame] dxcompiler.dll!llvm::DenseMapBase<llvm::DenseMap<llvm::StringRef,clang::spirv::SpirvString *,clang::spirv::StringMapInfo,llvm::detail::DenseMapPair<llvm::StringRef,clang::spirv::SpirvString *>>,llvm::StringRef,clang::spirv::SpirvString *,clang::spirv::StringMapInfo,llvm::detail::DenseMapPair<llvm::StringRef,clang::spirv::SpirvString *>>::InsertIntoBucket(llvm::StringRef && TheBucket, clang::spirv::SpirvString * &&) Line 419	C++
 	[Inline Frame] dxcompiler.dll!llvm::DenseMapBase<llvm::DenseMap<llvm::StringRef,clang::spirv::SpirvString *,clang::spirv::StringMapInfo,llvm::detail::DenseMapPair<llvm::StringRef,clang::spirv::SpirvString *>>,llvm::StringRef,clang::spirv::SpirvString *,clang::spirv::StringMapInfo,llvm::detail::DenseMapPair<llvm::StringRef,clang::spirv::SpirvString *>>::FindAndConstruct(llvm::StringRef &&) Line 242	C++
 	dxcompiler.dll!llvm::DenseMapBase<llvm::DenseMap<llvm::StringRef,clang::spirv::SpirvString *,clang::spirv::StringMapInfo,llvm::detail::DenseMapPair<llvm::StringRef,clang::spirv::SpirvString *>>,llvm::StringRef,clang::spirv::SpirvString *,clang::spirv::StringMapInfo,llvm::detail::DenseMapPair<llvm::StringRef,clang::spirv::SpirvString *>>::operator[](llvm::StringRef && Key) Line 246	C++
 	dxcompiler.dll!clang::spirv::SpirvBuilder::getString(llvm::StringRef str) Line 1645	C++
 	dxcompiler.dll!clang::spirv::DeclResultIdMapper::createOrUpdateStringVar(const clang::VarDecl * var) Line 1105	C++
 	dxcompiler.dll!clang::spirv::SpirvEmitter::doVarDecl(const clang::VarDecl * decl) Line 1556	C++
 	dxcompiler.dll!clang::spirv::SpirvEmitter::doDecl(const clang::Decl * decl) Line 813	C++
 	dxcompiler.dll!clang::spirv::SpirvEmitter::HandleTranslationUnit(clang::ASTContext & context) Line 661	C++
 	dxcompiler.dll!clang::ParseAST(clang::Sema & S, bool PrintStats, bool SkipFunctionBodies) Line 162	C++
 	dxcompiler.dll!clang::ASTFrontendAction::ExecuteAction() Line 556	C++
 	dxcompiler.dll!clang::FrontendAction::Execute() Line 469	C++
 	dxcompiler.dll!DxcCompiler::Compile(const DxcBuffer * pSource, const wchar_t * * pArguments, unsigned int argCount, IDxcIncludeHandler * pIncludeHandler, const _GUID & riid, void * * ppResult) Line 1031	C++
```

### PR DESCRIPTION
**Pull request by greg-lunarg**
Still need to add test.

Fixes #4360

### CODE REVIEW

**Comment by jaebaek** [comment]
This looks good to me. I am waiting for the unit tests.

File: tools/clang/lib/SPIRV/SpirvBuilder.cpp [hunk 1]
```diff
@@ -1666,14 +1666,22 @@ SpirvConstant *SpirvBuilder::getConstantNull(QualType type) {
 
 SpirvString *SpirvBuilder::getString(llvm::StringRef str) {
```
Comment (jaebaek) [inline tools/clang/lib/SPIRV/SpirvBuilder.cpp:1667]: Nit: This is minor issue, but now this method becomes long .. changing the method name to `SpirvBuilder::getOrCreateString(..)` would be nice.
In addition, I'd like to separate the part to create the string to another method i.e.,
```
SpirvString *SpirvBuilder::createString(llvm::StringRef str) {
  // Create a SpirvString instruction
  auto *instr = new (context) SpirvString(/* SourceLocation */ {}, str);
  instr->setRValue();
  stringLiterals[str.str()] = instr;
  if (str.empty())
    emptyString = instr;
  else
    stringLiterals[str.str()] = instr;
  mod->addString(instr);
  return instr;
}
```

Comment (greg-lunarg) [reply tools/clang/lib/SPIRV/SpirvBuilder.cpp:1723]: Done.

**Review by jaebaek — APPROVED** [review]

**Comment by jaebaek** [comment]
LGTM. If you think it is fine to submit, I will do it. Please let me know.

**Comment by greg-lunarg** [comment]
@jaebaek Thanks! Please merge.

**Pull request merged**